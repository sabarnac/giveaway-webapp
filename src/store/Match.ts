import { observable, computed, action } from "mobx";
import Participant from "./Participant";
import Random from "random-js";
import Config from "../config/Config";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _participants: Participant[];
  @observable private _winner: Participant;
  private static _RANDOM_GENERATOR: Random = Config.randomGenerator;

  private _getWinner = (): Participant =>
    Match._RANDOM_GENERATOR.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  public get winner(): Participant {
    return this._winner;
  }
}
