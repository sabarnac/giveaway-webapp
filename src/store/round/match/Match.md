Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const match: Match = new Match(
  [new Participant("FOO"), new Participant("BAR")],
  "1",
);
const matchId: string = match.id; // "1"
const matchCompleteId: string = match.fullId; // "1:1"
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
const winner: Participant[] = round.winner; // Participant("FOO")
```
