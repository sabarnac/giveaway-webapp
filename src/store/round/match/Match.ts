import { observable, computed } from "mobx";
import Participant from "./participant/Participant";
import RandomGenerator from "../../config/RandomGenerator";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _id: string;
  @observable private _participants: Participant[];
  @observable private _winner: Participant;

  private static counter: number = 1;

  private _getWinner = (): Participant =>
    RandomGenerator.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._id = `match-${Match.counter++}`;
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  @computed public get winner(): Participant {
    return this._winner;
  }

  /**
   * Determines whether another match is identical to the current one.
   * @param  {Match} otherMatch The other match to compare against.
   * @returns {boolean} Whether the other match is equal to the current one.
   */
  public equals = (otherMatch: Match): boolean => this._id === otherMatch._id;
}
