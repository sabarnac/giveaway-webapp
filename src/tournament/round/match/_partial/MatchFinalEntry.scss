.match__interim {
  width: 50%;
  display: flex;
  align-items: center;
  & > .participant-entry {
    border-left: none;
  }
  opacity: 0.000001;
  will-change: opacity;
  &--exiting,
  &--exited {
    opacity: 0.000001;
  }
  &--entered,
  &--entering,
  &--completed {
    opacity: 0.99999;
  }
  & > .participant {
    background: #ffaa00;
  }
}
