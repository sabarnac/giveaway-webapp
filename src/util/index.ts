export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
