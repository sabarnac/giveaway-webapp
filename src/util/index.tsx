import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
