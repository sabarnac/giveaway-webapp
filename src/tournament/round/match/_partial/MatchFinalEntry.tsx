import React, { useEffect, Fragment } from "react";
import "./MatchFinalEntry.scss";
import classNames from "classnames";
import ParticipantEntry from "../participant/entry/ParticipantEntry";
import { getNormalizedSpeed, isInRange } from "../../../../util/index";
import { CSSTransition } from "react-transition-group";
import {
  useAnimationState,
  AnimationStateHook,
  runOnPredicate,
} from "../../../../util";
import Match from "../../../../store/round/match/Match";
import Participant from "../../../../store/round/match/participant/Participant";
import MatchWinnerView from "./MatchWinnerView";

/**
 * Properties of the match final entry React component.
 */
interface MatchFinalEntryProps {
  /** CSS class */
  className: string;
  /** Whether it is the currently ongoing match. */
  isCurrentMatch: boolean;
  /** Whether it is an actual match. */
  isActualMatch: boolean;
  /** The details of the current match. */
  match: Match;
}

/**
 * React component for the match final entry.
 */
export default (props: MatchFinalEntryProps): JSX.Element => {
  const [
    currentState,
    updateState,
    updateStateDelay,
    setState,
  ]: AnimationStateHook = useAnimationState();

  useEffect(runOnPredicate(currentState === 0, updateState));
  useEffect(
    runOnPredicate((currentState + 1) % 3 === 0, () =>
      updateStateDelay(getNormalizedSpeed(200)),
    ),
  );
  useEffect(
    runOnPredicate(
      currentState === props.match.participants.length * 3 + 1,
      () => setState(1),
    ),
  );

  return (
    <Fragment>
      {props.isCurrentMatch && props.isActualMatch ? (
        props.match.participants.map(
          (participant: Participant, index: number): JSX.Element => (
            <CSSTransition
              key={participant.id}
              in={isInRange(currentState, index * 3 + 1, index * 3 + 2)}
              timeout={getNormalizedSpeed(200)}
              classNames={{
                enter: "",
                enterActive: `${props.className}__interim--entering`,
                enterDone: `${props.className}__interim--entered`,
                exit: "",
                exitActive: `${props.className}__interim--exiting`,
                exitDone: `${props.className}__interim--exited`,
              }}
              mountOnEnter={true}
              unmountOnExit={true}
              onEntered={updateState}
              onExited={updateState}
            >
              <div
                className={classNames(`${props.className}__interim`)}
                style={{
                  transition: `opacity ${getNormalizedSpeed(
                    200,
                  )}ms ease-in-out`,
                }}
              >
                <ParticipantEntry participant={participant} />
              </div>
            </CSSTransition>
          ),
        )
      ) : (
        <MatchWinnerView
          className={props.className}
          isCurrentMatch={props.isCurrentMatch}
          winner={props.match.winner}
        />
      )}
    </Fragment>
  );
};
