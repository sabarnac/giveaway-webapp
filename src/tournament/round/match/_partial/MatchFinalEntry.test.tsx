import React from "react";
import MatchFinalEntry from "./MatchFinalEntry.react";
import Match from "../../../../store/round/match/Match";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyMatch } from "../../../../util/test";

describe("MatchFinalEntry Component.", () => {
  const match: Match = createDummyMatch();
  describe("Current And Actual Match.", () => {
    it("Matches snapshot.", () => {
      const component: ReactTestRenderer = create(
        <MatchFinalEntry
          match={match}
          className="foobar"
          isCurrentMatch={true}
          isActualMatch={true}
        />,
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
  describe("Not Current But Actual Match.", () => {
    it("Matches snapshot.", () => {
      const component: ReactTestRenderer = create(
        <MatchFinalEntry
          match={match}
          className="foobar"
          isCurrentMatch={false}
          isActualMatch={true}
        />,
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
  describe("Current But Not Actual Match.", () => {
    it("Matches snapshot.", () => {
      const component: ReactTestRenderer = create(
        <MatchFinalEntry
          match={match}
          className="foobar"
          isCurrentMatch={true}
          isActualMatch={false}
        />,
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
  describe("Not Current Nor Actual Match.", () => {
    it("Matches snapshot.", () => {
      const component: ReactTestRenderer = create(
        <MatchFinalEntry
          match={match}
          className="foobar"
          isCurrentMatch={false}
          isActualMatch={false}
        />,
      );

      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
