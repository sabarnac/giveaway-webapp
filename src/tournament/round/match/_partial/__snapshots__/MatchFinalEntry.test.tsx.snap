// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchFinalEntry Component. Current And Actual Match. Matches snapshot. 1`] = `
Array [
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__interim--entering",
        "enterDone": "foobar__interim--entered",
        "exit": "",
        "exitActive": "foobar__interim--exiting",
        "exitDone": "foobar__interim--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__interim"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-entry"
      >
        <div
          className="participant-entry__avatar"
        >
          <img
            alt="barfoo-1"
            src="foobar-1"
          />
        </div>
        <div
          className="participant-entry__name"
        >
          <strong>
            Fozbaz 1
          </strong>
        </div>
      </div>
    </div>
  </div>,
  <div
    classNames={
      Object {
        "enter": "",
        "enterActive": "foobar__interim--entering",
        "enterDone": "foobar__interim--entered",
        "exit": "",
        "exitActive": "foobar__interim--exiting",
        "exitDone": "foobar__interim--exited",
      }
    }
    in={false}
    mountOnEnter={true}
    onEntered={[Function]}
    onExited={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="foobar__interim"
      style={
        Object {
          "transition": "opacity 200ms ease-in-out",
        }
      }
    >
      <div
        className="participant participant-entry"
      >
        <div
          className="participant-entry__avatar"
        >
          <img
            alt="barfoo-2"
            src="foobar-2"
          />
        </div>
        <div
          className="participant-entry__name"
        >
          <strong>
            Fozbaz 2
          </strong>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`MatchFinalEntry Component. Current But Not Actual Match. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "foobar__winner--entering",
      "enterDone": "foobar__winner--entered",
      "exit": "",
      "exitActive": "foobar__winner--exiting",
      "exitDone": "foobar__winner--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  timeout={200}
  unmountOnExit={true}
>
  <div
    className="foobar__winner"
    style={
      Object {
        "transition": "opacity 200ms ease-in-out",
      }
    }
  >
    <div
      className="participant participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          Fozbaz 1
        </strong>
      </div>
    </div>
  </div>
</div>
`;

exports[`MatchFinalEntry Component. Not Current But Actual Match. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "foobar__winner--entering",
      "enterDone": "foobar__winner--entered",
      "exit": "",
      "exitActive": "foobar__winner--exiting",
      "exitDone": "foobar__winner--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  timeout={200}
  unmountOnExit={true}
>
  <div
    className="foobar__winner"
    style={
      Object {
        "transition": "opacity 200ms ease-in-out",
      }
    }
  >
    <div
      className="participant participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          Fozbaz 1
        </strong>
      </div>
    </div>
  </div>
</div>
`;

exports[`MatchFinalEntry Component. Not Current Nor Actual Match. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "foobar__winner--entering",
      "enterDone": "foobar__winner--entered",
      "exit": "",
      "exitActive": "foobar__winner--exiting",
      "exitDone": "foobar__winner--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  timeout={200}
  unmountOnExit={true}
>
  <div
    className="foobar__winner"
    style={
      Object {
        "transition": "opacity 200ms ease-in-out",
      }
    }
  >
    <div
      className="participant participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="barfoo-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          Fozbaz 1
        </strong>
      </div>
    </div>
  </div>
</div>
`;
